src/details/ArborX_DetailsTreeTraversal.hpp

```diff
@@ -450,6 +450,146 @@ struct TreeTraversal<BVH, Predicates, Callback, NearestPredicateTag>
+struct TreeTraversal<BVH, Predicates, Callback,
+                     Experimental::OrderedNearestPredicateTag>
+{
+  BVH _bvh;
+  Predicates _predicates;
+  Callback _callback;
+
+  using Access = AccessTraits<Predicates, PredicatesTag>;
+
+  template <class ExecutionSpace>
+  TreeTraversal(ExecutionSpace const &space, BVH const &bvh,
+                Predicates const &predicates, Callback const &callback)
+      : _bvh{bvh}
+      , _predicates{predicates}
+      , _callback{callback}
+  {
+    if (_bvh.empty())
+    {
+      // do nothing
+    }
+    else if (_bvh.size() == 1)
+    {
+      Kokkos::parallel_for(
+          "ArborX::Experimental::TreeTraversal::OrderedNearestPredicate"
+          "degenerated_one_leaf_tree",
+          Kokkos::RangePolicy<ExecutionSpace, OneLeafTree>(
+              space, 0, Access::size(predicates)),
+          *this);
+    }
+    else
+    {
+      Kokkos::parallel_for(
+          "ArborX::Experimental::TreeTraversal::OrderedNearestPredicate",
+          Kokkos::RangePolicy<ExecutionSpace>(space, 0,
+                                              Access::size(predicates)),
+          *this);
+    }
+  }
+
+  struct OneLeafTree
+  {};
+
+  KOKKOS_FUNCTION void operator()(OneLeafTree, int queryIndex) const
+  {
+    auto const &predicate = Access::get(_predicates, queryIndex);
+    auto const root = HappyTreeFriends::getRoot(_bvh);
+    auto const &root_bounding_volume =
+        HappyTreeFriends::getBoundingVolume(_bvh, root);
+    constexpr auto inf = KokkosExt::ArithmeticTraits::infinity<float>::value;
+    if (distance(getGeometry(predicate), root_bounding_volume) != inf)
+    {
+      _callback(predicate, 0);
+    }
+  }
+
+  KOKKOS_FUNCTION void operator()(int queryIndex) const
+  {
+    auto const &predicate = Access::get(_predicates, queryIndex);
+    using ArborX::Details::HappyTreeFriends;
+
+    auto const distance = [geometry = getGeometry(predicate),
+                           bvh = _bvh](int node) {
+      auto const &box = HappyTreeFriends::getBoundingVolume(bvh, node);
+      using Details::distance;
+      return distance(geometry, box);
+    };
+
+    using PairIndexDistance = Kokkos::pair<int, float>;
+    struct CompareDistance
+    {
+      KOKKOS_FUNCTION bool operator()(PairIndexDistance const &lhs,
+                                      PairIndexDistance const &rhs) const
+      {
+        return lhs.second > rhs.second;
+      }
+    };
+
+    PairIndexDistance heap[64];
+    PairIndexDistance *heap_last = heap;
+    CompareDistance const compare;
+    constexpr auto inf = KokkosExt::ArithmeticTraits::infinity<float>::value;
+
+    int node = HappyTreeFriends::getRoot(_bvh);
+    int left_child;
+    int right_child;
+
+    while (true)
+    {
+      if (HappyTreeFriends::isLeaf(_bvh, node))
+      {
+        if (invoke_callback_and_check_early_exit(
+                _callback, predicate,
+                HappyTreeFriends::getLeafPermutationIndex(_bvh, node)))
+          return;
```

Comment:
Guess it would still allow one to find just `k` nearest ones, while processing them in the distance closest order.

Reply:
Yes, the question is whether we want it to rather behave like a spatial predicate or like a nearest predicate. I tend to nearest which is why I choose `OrderedNearest` for now. 
Also, this would be one of the traversals raytracing algorithms commonly use (multi-hit).

Reply:
we do check predicates, every time we compare to infinity